utils/middleware/citizen_id.ts

```diff
@@ -0,0 +1,59 @@
+import { toError } from "fp-ts/lib/Either";
+import { fromEither, taskEither, tryCatch } from "fp-ts/lib/TaskEither";
+import { IRequestMiddleware } from "io-functions-commons/dist/src/utils/request_middleware";
+import {
+  IResponseErrorForbiddenNotAuthorized,
+  IResponseErrorValidation,
+  ResponseErrorForbiddenNotAuthorized,
+  ResponseErrorValidation
+} from "italia-ts-commons/lib/responses";
+import { FiscalCode, NonEmptyString } from "italia-ts-commons/lib/strings";
+import { CitizenID } from "../../generated/definitions/CitizenID";
+import { withCitizenIdCheck } from "../citizen_id";
+import { RequiredHeaderMiddleware } from "./required_header";
+
+/**
+ * Perform validation on an incoming request assuring that
+ * - a CitizenID header is provided in the correct format
+ * - the CitizenID is a valid JWT
+ * - the CitizenID value is resolved into a valid FiscalCode
+ *
+ * @param publicRsaCertificate RSA certificate to validate JWT signature
+ * @param headerName (optional) the header in which we expect the value, default: "x-citizen-id"
+ *
+ * @returns Either the resolved fiscal code or a proper error
+ */
+export const RequestCitizenToFiscalCode = (
+  publicRsaCertificate: NonEmptyString,
+  headerName = "x-citizen-id"
+): IRequestMiddleware<
+  "IResponseErrorValidation" | "IResponseErrorForbiddenNotAuthorized",
+  FiscalCode
+> => async request => {
+  return (
+    taskEither
+      .of<
+        IResponseErrorValidation | IResponseErrorForbiddenNotAuthorized,
+        void
+      >(void 0)
+      // validate the header is present and its value is in the correct shape
+      .chain(_ =>
+        tryCatch(
+          () => RequiredHeaderMiddleware(headerName, CitizenID)(request),
+          e =>
+            ResponseErrorValidation(
+              `Invalid ${headerName} header`,
+              toError(e).message
+            )
+        )
+      )
+      .chain(fromEither)
+      // validate the value is either a fiscal code or a valid JWT. In the latter case, the fiscal code is resolved
+      .chain(c =>
+        withCitizenIdCheck(c, publicRsaCertificate, fiscalCode =>
```

Comment:
I will add another check on ADB2C User's groups, in order to allow direct fiscalCode usage by admin users. In this case I' ll add other input params and last but not least, the AdUser's `oid`. I' m wondering if we can let this code into a middleware, because I'm not sure we can access the AdUser at this level (there's another middleware that extracts ADUser from authentication header).